### ISSUE TITLE
Combining model and data parallelism

### ISSUE THREAD

**Issue by dptam** [pytorch/xla#2677]
## ❓ Questions and Help
Hi, 

I understand how to do model parallelism and data parallelism, and was wondering how to combine model and data parallelism. The model parallelism only has be done once (across the whole batch on all cores, not per core). In particular, if x is the batch and a is some data independent of the batch size, the loss if f(x) * g(a). Scaling the number of cores can scale computing f(x), but g(a) only has be computed once on 1 core. I was wondering how I would implement this? 

In particular, I was thinking g(a) would first be computed on 1 core, and then it would be transferred to all the other cores so they can compute f(x) * g(a)? But the processes computing f(x) * g(a) would have to wait for g(a) to be computed first on a separate process?  

Thanks

**Comment by taylanbil** [comment]
so, let me see if I understand your case:

1. f(x) is computed in a data-parallel way. you shard the batches and process data parallel.
1. g(a) is a totally independent computation, has nothing to do w/ f(x).

Since you're suggesting g(a) is computable on 1 core, I'm assuming that's a quick, lightweight computation. In that case, why not do that exact computation on all cores?

I don't think you can trivially reserve 1 process for a completely different computation, and have `worldsize-1` processes do data parallelism. Am I misunderstanding your question?

**Comment by dptam** [comment]
Thanks for the quick reply. 

g(a) is an expensive operation that requires the same memory/compute as f(x). I don't follow the worldsize-1 processes doing data parallelism? 

I was thinking with `worldsize-8`, then it would be 
 if `rank(process) == 1 `, then `compute g(a)`, else `compute L = g(a)*f(x); backprop(L)`

Processes greater than 1 would do data parallelism, while process 1 would do model parallelism with all the processes greater than 1.

**Comment by taylanbil** [comment]
It's possible to do different work on cores, have model-parallel and data-parallel groups etc, but having 1 core do something completely different is very tricky, potentially problematic w/ the current xla model.

* is it possible to shard the computation of g(a) at all? If so, you could compute g(a) 8-way data parallel, and then compute f(x) data parallel as well, presumably this would not be sub-optimal (it's not wasting a compute). this is likely the easiest solution.
* it may be possible to compute g(a) on one device, f(x) on 7 other devices, and then do all_gather's to collect f1(x) ... f7(x) and g(a) on all 8 cores. This requires you to know the shape of f(x) and g(a) beforehand, so you can use a dummy tensor for g(a) on ordinals 1...7 etc. And then you'd need to handle backprop very carefully. we can't do a 7-way all reduce on f's optimizer, so you're going to need to have a dummy optimizer in ordinal 0, w/ 0 grads, and do a full allreduce, and then adjust your weights (*7/8 or something). g's backprop don't need an allreduce so you're fine there.

Tbh I'm not sure if the second approach would work, but if it did, you'd be parallelizing the computation of g(a) and f(x).

**Comment by taylanbil** [comment]
@dptam Is this issue good to close?

**Comment by dptam** [comment]
Yea, thanks for the suggestions. I will try suggestion 1.
